module/zfs/spa.c

```diff
@@ -1728,13 +1728,15 @@ spa_config_valid(spa_t *spa, nvlist_t *config)
 	vdev_t *mrvd, *rvd = spa->spa_root_vdev;
 	nvlist_t *nv;
 	int c, i;
+	boolean_t is_valid = B_FALSE;
 
 	VERIFY(nvlist_lookup_nvlist(config, ZPOOL_CONFIG_VDEV_TREE, &nv) == 0);
 
 	spa_config_enter(spa, SCL_ALL, FTAG, RW_WRITER);
 	VERIFY(spa_config_parse(spa, &mrvd, nv, NULL, 0, VDEV_ALLOC_LOAD) == 0);
 
-	ASSERT3U(rvd->vdev_children, ==, mrvd->vdev_children);
+	if (rvd->vdev_children != mrvd->vdev_children)
+		goto cleanup;
```

Comment:
LGTM.  In this case it looks like the best thing to do is fail.  However, as a general rule we should try and handle as much as possible in the import path.